File: Heap-Priority_Queue/Easy/LC_1046-Last_Stone_Weight/LC_1046-Last_Stone_Weight.py
import heapq

# Heap/Priority Queue approach
# T: O(n log n), M: O(1)
# Where n is the length of stones list

class Solution:
    def lastStoneWeight(self, stones: list[int]) -> int:
        # Create stones max heap
        stones = [-weight for weight in stones]
        heapq.heapify(stones)
        # Smash stones until there is 1 stone left
        while len(stones) > 1:
            heaviest1, heaviest2 = heapq.heappop(stones), heapq.heappop(stones)
            heapq.heappush(stones, heaviest1 - heaviest2)
        return -stones[0]

# Ex1:
attempt = Solution().lastStoneWeight(stones = [2,7,4,1,8,1])
assert attempt == 1, f'Expected 1, but got {attempt}'
# Ex2:
attempt = Solution().lastStoneWeight(stones = [1])
assert attempt == 1, f'Expected 1, but got {attempt}'
# Test case 3:
attempt = Solution().lastStoneWeight(stones = [1, 1])
assert attempt == 0, f'Expected 0, but got {attempt}'

File: Heap-Priority_Queue/Easy/LC_1046-Last_Stone_Weight/LC_1046-Last_Stone_Weight.cpp
#include <bits/stdc++.h>

using namespace std;

// Heap/Priority Queue approach
// T: O(n log n), M: O(1)
// Where n is the length of stones list

class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        // Create stones max priority queue
        priority_queue<int> pq = priority_queue<int>(stones.begin(), stones.end());
        // Smash stones until there is 1 stone left
        while (pq.size() > 1) {
            int heaviest1 = pq.top(); pq.pop();
            int heaviest2 = pq.top(); pq.pop();
            pq.push(heaviest1 - heaviest2);
        }
        return pq.top();
    }
};

int main() {
    vector<int> stones;
    int attempt, answer;

    // Example 1:
    stones = {2,7,4,1,8,1};
    attempt = Solution().lastStoneWeight(stones);
    answer = 1;
    assert(attempt == answer);
    // Example 2:
    stones = {1};
    attempt = Solution().lastStoneWeight(stones);
    answer = 1;
    assert(attempt == answer);
    // Test case 3:
    stones = {1, 1};
    attempt = Solution().lastStoneWeight(stones);
    answer = 0;
    assert(attempt == answer);
}